openmls/src/ciphersuite/mod.rs

```diff
@@ -546,63 +549,150 @@ impl Signature {
-// - 1 byte length tag
-// - 1 byte to indicate if the integer is signed
-const MAX_DER_OVERHEAD: usize = 9;
+// DER encoding SEQUENCE tag.
+const SEQUENCE_TAG: u8 = 0x30;
 
 impl Signature {
     /// This function takes a DER encoded ECDSA signature and decodes it to the
     /// bytes representing the concatenated scalars. If the decoding fails, it
     /// will throw a `der::Error`.
-    pub(crate) fn der_decode(&self) -> Result<Vec<u8>, der::Error> {
-        // A small closure to decode a single scalar and pad it with leading
-        // zeroes until it's `SCALAR_SIZE` bytes long.
-        let decode_scalar = |decoder: &mut der::Decoder| -> Result<Vec<u8>, der::Error> {
-            let decoded_scalar = UIntBytes::decode(decoder)?;
-            let decoded_scalar_len: usize = decoded_scalar.len().try_into()?;
+    pub(crate) fn der_decode(&self) -> Result<Vec<u8>, SignatureError> {
+        // A small function to DER decode a single scalar.
+        fn decode_scalar<R: Read>(mut buffer: R) -> Result<Vec<u8>, SignatureError> {
+            // Check header bytes of encoded scalar.
+
+            // 1 byte INTEGER tag should be 0x02
+            let integer_tag = buffer.read_u8()?;
+            if integer_tag != INTEGER_TAG {
+                return Err(SignatureError::DecodingError);
+            };
+
+            // 1 byte length tag should be at most 0x21, i.e. 32 plus at most 1
+            // byte indicating that the integer is unsigned.
+            let mut scalar_length = buffer.read_u8()? as usize;
+            if scalar_length > P256_SCALAR_LENGTH + 1 {
+                return Err(SignatureError::DecodingError);
+            };
+
+            // If the scalar is 0x21 long, the first byte has to be 0x00,
+            // indicating that the following integer is unsigned. We can discard
+            // this byte safely. If it's not 0x00, the scalar is too large not
+            // thus not a valid point on the curve.
+            if scalar_length == P256_SCALAR_LENGTH + 1 {
+                if buffer.read_u8()? != 0x00 {
+                    return Err(SignatureError::DecodingError);
+                };
+                // Since we just read that byte, we decrease the length by 1.
+                scalar_length -= 1;
+            };
+
+            let mut scalar = vec![0; scalar_length];
+            buffer.read_exact(&mut scalar)?;
+
+            // The scalar should have the right length.
+            if scalar.len() != scalar_length {
+                return Err(SignatureError::DecodingError);
+            };
+
             // The verification algorithm expects the scalars to be 32 bytes
             // long, buffered with zeroes.
-            let mut padded_scalar = vec![0u8; SCALAR_SIZE - decoded_scalar_len];
-            padded_scalar.extend_from_slice(decoded_scalar.as_bytes());
+            let mut padded_scalar = vec![0u8; P256_SCALAR_LENGTH - scalar_length];
+            padded_scalar.extend_from_slice(&scalar);
```

Comment:
You can avoid copying things around by reading into the `padded_scalar` directly. But then you have to check the length again. You can use `append` to make this a move rather than a copy, which would be fine I guess.